### PR TITLE
feat(contract): add buff if card on its role or captain

### LINKED ISSUE #24: Implement card buff if it's placed in its role
# Description

Each card has a specific role, if a card is placed on the corresponding role on the field it has a +1/+1 buff. For example if a 3/4 goalkeeper is placed on the goalkeeper slot it'll be buffed to 4/5.

# Acceptance criteria

* A card is buffed if it's played on it's role slot.

### PR DESCRIPTION
**Pull request by Matth26**
Fix #24

I added the buff if card is captain as well.

### CODE REVIEW

File: contracts/src/systems/place_card.cairo [hunk 1]
```diff
@@ -23,27 +23,72 @@ mod place_card_system {
...
         }
+
+        if (is_on_its_role) {
+            card.current_dribble += 1;
+            card.current_defense += 1;
+        }
+        if (card.is_captain) {
+            card.current_dribble += 1;
+            card.current_defense += 1;
+        }
```
Comment (LucasLvy) [inline contracts/src/systems/place_card.cairo:88]:
```suggestion
        if is_on_its_role || card.is_captain {
            card.current_dribble += 1;
            card.current_defense += 1;
        }
```

File: contracts/src/systems/place_card.cairo [hunk 2]
```diff
@@ -23,27 +23,72 @@ mod place_card_system {
...
+                match card.role {
+                    Roles::Goalkeeper => (),
+                    Roles::Defender => (),
+                    Roles::Midfielder => (),
+                    Roles::Attacker => {
+                        is_on_its_role = true;
+                    },
+                }
             }
         }
```
Comment (LucasLvy) [inline contracts/src/systems/place_card.cairo:79]: WDYT of that ?
```rs
        let is_on_its_role = match position {
              Roles::Goalkeeper => {
                  assert(player.goalkeeper.is_none(), 'Goalkeeper already placed');
                  player.goalkeeper = Option::Some(Placement::Side(card_id));
                card.role == Roles::Goalkeeper
              },
              Roles::Defender => {
                  assert(player.defender.is_none(), 'Defender already placed');
                  player.defender = Option::Some(Placement::Side(card_id));
                card.role == Roles::Defender
              },
              Roles::Midfielder => {
                  assert(player.midfielder.is_none(), 'Midfielder already placed');
                  player.midfielder = Option::Some(Placement::Side(card_id));
                card.role == Roles::Midfielder
              },
              Roles::Attacker => {
                  assert(player.attacker.is_none(), 'Attacker already placed');
                  player.attacker = Option::Some(Placement::Side(card_id));
                card.role == Roles::Attacker
              }
        };
```

File: contracts/src/tests/place_card.cairo [hunk 3]
```diff
@@ -79,3 +79,139 @@ fn test_place_card_overflow() {
...
+        role: Roles::Attacker,
+        is_captain: false
+    };
+    set!(world, (card));
+    set_caller_address(player1);
+    let mut place_card_calldata: Array = Default::default();
+    place_card_calldata.append(game_id);
+    place_card_calldata.append(1); // card_id.low
+    place_card_calldata.append(0); // card_id.high
+    place_card_calldata.append(0); // Roles::Goalkeeper
```
Comment (LucasLvy) [inline contracts/src/tests/place_card.cairo:139]:
```suggestion
    /// card_id.low, card_id.high, Roles::Goalkeeper
    let place_card_calldata = array![game_id, 1, 0, 0];
```

File: contracts/src/tests/place_card.cairo [hunk 4]
```diff
@@ -79,3 +79,139 @@ fn test_place_card_overflow() {
...
```
Comment (LucasLvy) [inline contracts/src/tests/place_card.cairo:173]: same here

File: contracts/src/tests/place_card.cairo [hunk 5]
```diff
@@ -79,3 +79,139 @@ fn test_place_card_overflow() {
...
+        role: Roles::Attacker,
+        is_captain: true
+    };
+    set!(world, (card));
+    set_caller_address(player1);
+    let mut place_card_calldata: Array = Default::default();
+    place_card_calldata.append(game_id);
+    place_card_calldata.append(1); // card_id.low
+    place_card_calldata.append(0); // card_id.high
+    place_card_calldata.append(0); // Roles::Goalkeeper
```
Comment (LucasLvy) [inline contracts/src/tests/place_card.cairo:207]: same here

File: contracts/src/tests/place_card.cairo [hunk 6]
```diff
@@ -79,3 +79,139 @@ fn test_place_card_overflow() {
...
+        role: Roles::Attacker,
+        is_captain: false
+    };
+    set!(world, (card));
+    set_caller_address(player1);
+    let mut place_card_calldata: Array = Default::default();
+    place_card_calldata.append(game_id);
+    place_card_calldata.append(1); // card_id.low
+    place_card_calldata.append(0); // card_id.high
+    place_card_calldata.append(3); // Roles::Attacker
```
Comment (LucasLvy) [inline contracts/src/tests/place_card.cairo:105]: same here

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: contracts/src/systems/place_card.cairo (hunk 2 above)
Comment (Matth26) [reply contracts/src/systems/place_card.cairo:79]: Waaaaayy better hahaha. I encountered a compiler error and concluded it was from the equality test. I thought it was a Cairo limitation with structs not implementing PartialEq, which is why I turned to the match statement.

File: contracts/src/systems/place_card.cairo (hunk 1 above)
Comment (LucasLvy) [reply contracts/src/systems/place_card.cairo]: you're right forget about that

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Pull request merged**